src/aws-cpp-sdk-core/include/smithy/client/features/Checksums.h

```diff
@@ -0,0 +1,245 @@
+            using HashingUtils = Aws::Utils::HashingUtils;
+            using MD5 = Aws::Utils::Crypto::MD5;
+            using CRC32 = Aws::Utils::Crypto::CRC32;
+            using CRC32C = Aws::Utils::Crypto::CRC32C;
+            using Sha256 = Aws::Utils::Crypto::Sha256;
+            using Sha1 = Aws::Utils::Crypto::Sha1;
+            using PrecalculatedHash = Aws::Utils::Crypto::PrecalculatedHash;
+
+
+            static std::shared_ptr<Aws::IOStream> GetBodyStream(const Aws::AmazonWebServiceRequest& request)
+            {
+                if (request.GetBody() != nullptr)
+                {
+                    return request.GetBody();
+                }
+                // Return an empty string stream for no body
+                return Aws::MakeShared<Aws::StringStream>(AWS_SMITHY_CLIENT_CHECKSUM, "");
+            }
+
+            static void AddChecksumToRequest(const std::shared_ptr<Aws::Http::HttpRequest>& httpRequest,
+                                             const Aws::AmazonWebServiceRequest& request)
+            {
+                Aws::String checksumAlgorithmName = Aws::Utils::StringUtils::ToLower(
+                    request.GetChecksumAlgorithmName().c_str());
+                if (request.GetServiceSpecificParameters())
+                {
+                    auto requestChecksumOverride = request.GetServiceSpecificParameters()->parameterMap.find(
+                        "overrideChecksum");
+                    if (requestChecksumOverride != request.GetServiceSpecificParameters()->parameterMap.end())
+                    {
+                        checksumAlgorithmName = requestChecksumOverride->second;
+                    }
+                }
+
+                bool shouldSkipChecksum = request.GetServiceSpecificParameters() &&
+                    request.GetServiceSpecificParameters()->parameterMap.find("overrideChecksumDisable") !=
+                    request.GetServiceSpecificParameters()->parameterMap.end();
+
+                //Check if user has provided the checksum algorithm
+                if (!checksumAlgorithmName.empty() && !shouldSkipChecksum)
+                {
+                    // Check if user has provided a checksum value for the specified algorithm
+                    const Aws::String checksumType = "x-amz-checksum-" + checksumAlgorithmName;
+                    const HeaderValueCollection& headers = request.GetHeaders();
+                    const auto checksumHeader = headers.find(checksumType);
+                    bool checksumValueAndAlgorithmProvided = checksumHeader != headers.end();
+
+                    // For non-streaming payload, the resolved checksum location is always header.
+                    // For streaming payload, the resolved checksum location depends on whether it is an unsigned payload, we let AwsAuthSigner decide it.
+                    if (request.IsStreaming() && checksumValueAndAlgorithmProvided)
```

Comment:
not in this pr, but now that we have it factored our we should likely refactor this into something logically easier to follow. but lumping that change into this one seems reckless at best. thinking out loud, but nice we can refactor this easier in the future.